### ISSUE TITLE
"expected to eventually match, got <true>" while <true> was indeed expected

### ISSUE THREAD

**Issue by ldiqual** [Quick/Nimble#293]
It happens pretty randomly, I have this piece of test that triggers a very conflicting error:

```
expect { delegate.wasDidUpdateTimeSlotsCalled }.toEventually(beTrue())
```

![image](https://cloud.githubusercontent.com/assets/941453/15134874/d25a31b6-1624-11e6-86bd-20f931f616f2.png)

How can this assertion fail if `wasDidUpdateTimeSlotsCalled` is indeed true?

Nimble 4.0.0

**Comment by ashfurrow** [comment]
Hmm, puzzling. I've not seen that before. Have you made any progress on this? Since `expect` uses autoclosures, I would typically write this as

```swift
expect(delegate.wasDidUpdateTimeSlotsCalled).toEventually( beTrue() )
```

But I doubt that has any affect on the outcome.

**Comment by ldiqual** [comment]
Weird indeed. I'll update my code to use implicit closures and will close the issue if I can't reproduce within a week. Thanks for the followup!

**Comment by Blackjacx** [comment]
I have this issue too with nimble 4.1.0: 

`expect(panelMock.isSearchPanelAnimationInProgress).toEventually(beFalse())`

gives me 

`expected to eventually match, got <false>`

**Comment by jeffh** [comment]
Hey @Blackjacx,

Nimble's async polling is similar to your code with a couple of notable differences to cover more cases:

- Any stops of NSRunLoop will not cause the NSRunLoop to stop running. Some Apple internal frameworks will call stop that can cause a premature failure.
- There's additional verification that the match runs within a given time interval. This can be backlogged due to amount of work on the run loop (eg - `sleep(10)`).
- Nested runs of NSRunLoop can create non-deterministic failures of the outer run.
- Nimble adds a high-priority timeout in an attempt interrupt to the run loop before other work gets processed.

It's based on the behavior of XCTest's expectations, but only assumes work that can be done in-process. Nimble can't restart the process with saved test running state when the timeout triggers (Which is what Xcode's XCTest does sometimes).


--------

@ldiqual, What's the context of the code that changes the `wasDidUpdateTimeSlotsCalled` property? Is it on another thread? Does the code that sets it to `true` run by attaching to the run loop?

**Comment by Blackjacx** [comment]
Well thanks, i have another issue that I need to solve:

    if !isExecuted {
      isExecuted = true
      let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(0.5 * Double(NSEC_PER_SEC)))
      dispatch_after(delayTime, dispatch_get_main_queue()) {
        self.centerMapOnUserLocation()
      }
    }

Now when I test centerMapOnUserLocation using toEventually the dispatched task is blocked by the toEventually on the main thread.

Do you have any solution for that?

**Comment by ldiqual** [comment]
@jeffh This error still happens very infrequently, but I don't have reproduction steps so I'll close the issue. Thanks for following up!
